New nodes

cchintada-cng opened the pull request:
All nodes are added. Some of them are not tested with web application. Will continue on testing them.

rob-cng commented on src/wait-config.html:
```diff
@@ -0,0 +1,60 @@
+        }
+    })
+</script>
+
```
What does "wait config" mean? What is this node used for, vs the "explicit wait" node?

cchintada-cng replied:
Wait-config is used at the beginning of session to set the global waits for implicit wait for each element to be displayed, Page load wait and script waits. Explicit waits are used specifically for web elements to be displayed/enabled/selected/not displayed.. Explicit waits are used in a flow where we need to add some extra time to wait before we do actions on elements. Web pages with AJAX code will render page/iframe after certain time, So we can wait till the element is displayed and click.

rob-cng replied:
Would it be better to call this "implicit-wait-config" then?

cchintada-cng replied:
Sure Rob! Will add the change

rob-cng commented on src/execute-script.html:
```diff
@@ -0,0 +1,54 @@
+        }
+    })
+</script>
+
```
Why does the execute script node need to locate an element? Doesn't it just execute javascript in the browser?

cchintada-cng replied:
It does! But it also find the element and pass element object as argument. which was java/C# selenium pattern!

rob-cng commented on src/element-check.html:
```diff
@@ -0,0 +1,51 @@
+
+    <div class="form-row">
+        <label for="node-input-check"><i class="fa fa-tasks"></i> Check</label>
+        <select type="text" id="node-input-check" style="width:70%;" onchange="setAction()">
```
This refers to `setAction()` but that function is not declared above

cchintada-cng replied:
removed onChange.

rob-cng commented on src/alert-action.js:
```diff
@@ -0,0 +1,50 @@
+               
+                let value = config.sendText || msg.value
+                
+                let isAlertPresent = async () => {
```
This is an independent function so should be moved outside the `alertAction` function. Could be declared before or after `alertAction`, something like `async function isAlertPresent() {`

cchintada-cng replied:
onst isAlertPresent = async (browser) => {
    try{
        await browser.getAlertText()
        return true
    }
    catch(e){
        return false
    } 
}

module.exports = function(RED) {

    function alertAction(config) {
        RED.nodes.createNode(this, config)
        const node = this        
        common.clearStatus(node)
``

cchintada-cng replied:
Tested with latest changes.

rob-cng commented on src/alert-action.html:
```diff
@@ -0,0 +1,50 @@
+            <option value="accept">Accept Alert</option>
+            <option value="getAlertText">Get Alert Text</option>
+            <option value="sendAlertText">Send Alert Text</option>
+            <option value="isExists">Is Alert Exists</option>
```
Change to "Is Alert Present", and "isPresent"

cchintada-cng replied:
changed as suggested

rob-cng commented on src/alert-action.js:
```diff
@@ -0,0 +1,50 @@
+                        //await browser.keys(Array.from(value))
+                    }                    
+                }
+                else if (config.action === 'isExists') {                    
```
isPresent

cchintada-cng replied:
changes as suggested

rob-cng commented on src/element-action.html:
```diff
@@ -53,13 +58,21 @@
         <input id="node-input-sendKeys" type="text">
     </div>
-
+    <div class="form-row" id="get_Attribute">
```
For consistency, this should be named `getAttribute` instead of `get_Attribute` (and update your scripts above to reflect this change)

cchintada-cng replied:
changed as suggested

cchintada-cng commented on src/wdio-common.js:
```diff
@@ -2,7 +2,7 @@ const wdio = require ('webdriverio')
 
 module.exports.getBrowser = (context) => {
-    let browser = context.flow.get('wdio_browser')
+    let browser = context.global.get('wdio_browser')
```
Currently we are using each node-red flow for a web page (like page object modeling) and a node-red project for a work-flow, so we are using the same session across multiple flows. But we can discuss on this.

cchintada-cng commented on src/wdio-common.js:
```diff
@@ -69,6 +70,37 @@ module.exports.getElementId = async (browser, using, value) => {
+    }
+
+    try{
+        element = browser.$(selector)
```
Sure Rob! Will add this change

cchintada-cng commented:
- Somehow yarn link is not working for me! So I had to use npm link. Can you please help me.
- Updated the package version.
- We have tested all the functionalities. To test the send text on alert message, I have written a small html to prompt an alert with text box. So the send alert text in alert-action is not worked when I tested with the code. Except that everything is working.
- Pallavi/team requested to add some code to not close the session and browser when deploy new changes on flows. This is useful when we wanted to debug from middle of work flow, where we can directly interact with elements on browser rather than starting the flow from the beginning. Which is a value added. 

> We have added a check box on new-session to kiss session or keep open the session

.

rob-cng commented on src/new-session.html:
```diff
@@ -47,7 +52,10 @@
     </div>
-
+    <div class="form-row">
+        <label for="node-input-killSession"><i class="fa fa-tasks"></i> Kill Session on Deploy</label>
```
For clarity, maybe "Kill existing session on flow deploy"
